File: python/normOutliers/stats_normOutliers_zMethod.py
#!/usr/bin/env python
# coding: utf-8

# # COURSE: Master statistics and machine learning: Intuition, Math, code
# ##### COURSE URL: udemy.com/course/statsml_x/?couponCode=202304 
# ## SECTION: Data normalizations and outliers
# ### VIDEO: Z-score for outlier removal
# #### TEACHER: Mike X Cohen, sincxpress.com

# In[ ]:


import numpy as np
import matplotlib.pyplot as plt
from statsmodels import robust
import scipy.stats as stats


# In[ ]:


## create some data

N = 40
data = np.random.randn(N)
data[data<-1] = data[data<-1]+2
data[data>2] = data[data>2]**2; # try to force a few outliers
data = data*200 + 50  # change the scale for comparison with z

# convert to z
dataZ = (data-np.mean(data)) / np.std(data)


#### specify the z-score threshold
zscorethresh = 3



# plot the data
fig,ax = plt.subplots(2,1,figsize=(8,6))

ax[0].plot(data,'k^',markerfacecolor='w',markersize=12)
ax[0].set_xticks([])
ax[0].set_xlabel('Data index')
ax[0].set_ylabel('Orig. scale')

# then plot the zscores
ax[1].plot(dataZ,'k^',markerfacecolor='w',markersize=12)
ax[1].plot([0,N],[zscorethresh,zscorethresh],'r--')
ax[1].set_xlabel('Data index')
ax[1].set_ylabel('Z distance')
plt.show()


# In[ ]:


## identify outliers

# find 'em!
outliers = np.where(abs(dataZ)>zscorethresh)[0]

# and cross those out
ax[0].plot(outliers,data[outliers],'x',color='r',markersize=20)
ax[1].plot(outliers,dataZ[outliers],'x',color='r',markersize=20)

fig


# In[ ]:


## iterative method

# pick a lenient threshold just for illustration
zscorethresh = 2
dataZ = (data-np.mean(data)) / np.std(data)


colorz = 'brkm'
numiters = 0 # iteration counter
while True:
    
    # convert to z
    datamean = np.nanmean(dataZ)
    datastd  = np.nanstd(dataZ)
    dataZ = (dataZ-datamean) / datastd
    
    # find data values to remove
    toremove = dataZ>zscorethresh
    
    # break out of while loop if no points to remove
    if sum(toremove)==0:
        break
    else:
        # otherwise, mark the outliers in the plot
        plt.plot(np.where(toremove)[0],dataZ[toremove],'%sx'%colorz[numiters],markersize=12)
        dataZ[toremove] = np.nan
    
    # replot
    plt.plot(dataZ,'k^',markersize=12,markerfacecolor=colorz[numiters],label='iteration %g'%numiters)
    numiters = numiters + 1

plt.xticks([])
plt.ylabel('Z-score')
plt.xlabel('Data index')
plt.legend()
plt.show()

#### the data points to be removed
removeFromOriginal = np.where(np.isnan(dataZ))[0]
print(removeFromOriginal)


# In[ ]:


## modified Z for non-normal distributions

# compute modified z
dataMed = np.median(data)
dataMAD = robust.mad(data)

dataMz = stats.norm.ppf(.75)*(data-dataMed) / dataMAD


# plot the data
fig,ax = plt.subplots(2,1,figsize=(8,6))

ax[0].plot(data,'k^',markerfacecolor='w',markersize=12)
ax[0].set_xticks([])
ax[0].set_xlabel('Data index')
ax[0].set_ylabel('Orig. scale')

# then plot the zscores
ax[1].plot(dataMz,'k^',markerfacecolor='w',markersize=12)
ax[1].plot([0,N],[zscorethresh,zscorethresh],'r--')
ax[1].set_xlabel('Data index')
ax[1].set_ylabel('Median dev. units (Mz)')
plt.show()

File: MATLAB/normOutliers/stats_normOutliers_zMethod.m
%%
%   COURSE: Master statistics and machine learning: intuition, math, code										
%      URL: udemy.com/course/statsml_x/?couponCode=202304 
% 
%  SECTION: Data normalizations and outliers
%    VIDEO: Z-score for outlier removal
% 
%  TEACHER: Mike X Cohen, sincxpress.com
%

%%

% a clear MATLAB workspace is a clear mental workspace
close all; clear; clc

%% create some data

N = 40;
data = randn(N,1);
data(data<-1) = data(data<-1)+2;
data(data>2) = data(data>2).^2; % try to force a few outliers
data = data*200 + 50; % change the scale for comparison with z

% convert to z
dataZ = (data-mean(data)) ./ std(data);


%%% specify the z-score threshold
zscorethresh = 3;


% show the data!
figure(1), clf
subplot(211), hold on
plot(data,'k^','markerfacecolor','w','markersize',12);
set(gca,'xtick',[]), box off
ylabel('Orig. scale')

subplot(212), hold on
plot(dataZ,'k^','markerfacecolor','w','markersize',12);
plot(get(gca,'xlim'),[1 1]*zscorethresh,'--','color','r')
set(gca,'xtick',[]), box off
ylabel('Stdev. units (z)')

%% identify outliers

% find the outliers (note the abs)
outliers = find(abs(dataZ)>zscorethresh);

% identify them in the plot
subplot(211)
plot(outliers,data(outliers),'rx','linew',3,'markersize',20)

subplot(212)
plot(outliers,dataZ(outliers),'rx','linew',3,'markersize',20)

%% iterative method

% pick a lenient threshold just for illustration
zscorethresh = 2;


figure(2), clf, hold on
colorz = 'brkm';
leg = {}; % initialize empty legend
numiters = 1; % iteration counter
while true
    
    % convert to z
    datamean = mean(dataZ,'omitnan');
    datastd  = std(dataZ,'omitnan');
    dataZ = (dataZ-datamean) ./ datastd;
    
    % find data values to remove
    toremove = dataZ>zscorethresh;
    
    % break out of while loop if no points to remove
    if sum(toremove)==0
        break
    else
        % otherwise, mark the outliers in the plot
        plot(find(toremove),dataZ(toremove),[ colorz(numiters) 'x' ],'markersize',12,'HandleVisibility','off')
        dataZ(toremove) = NaN;
    end
    
    % replot
    plot(dataZ,'k^','markersize',12,'markerfacecolor',colorz(numiters))
    
    % update the legend
    leg{numiters} = [ 'iteration ' num2str(numiters) ];
    numiters = numiters + 1;
end

set(gca,'xtick',[])
ylabel('Z-score'), xlabel('Data index')
legend(leg)


%%% the data points to be removed
removeFromOriginal = find(isnan(dataZ));

%% modified Z for non-normal distributions

% compute modified z
dataMed = median(data);
dataMAD = mad(data,1); % IMPORTANT: note the second input

dataMz = norminv(.75)*(data-dataMed) ./ dataMAD;


% show the data!
figure(3), clf
subplot(211), hold on
plot(data,'k^','markerfacecolor','w','markersize',12);
set(gca,'xtick',[]), box off
ylabel('Orig. scale')

subplot(212), hold on
plot(dataMz,'k^','markerfacecolor','w','markersize',12);
plot(get(gca,'xlim'),[1 1]*zscorethresh,'--','color','r')
set(gca,'xtick',[]), box off
ylabel('Median dev. units (Mz)')

%% done.
